### PR TITLE
#26: Implement wiro rpc client (closes #26)

### LINKED ISSUE #26: Implement wiro rpc client
## requirements
Implement wiro rpc client.
Parsing the trait of a controller, the wiro rpc client should generate an http client that is compatible to wiro generated http server.

## specs

Given
```scala
trait A {
  @command
  def do(input: String)
}
```
The client should generate something function to perform

```
POST /A/do -d {
  input: "whatever"
}
```

### PR DESCRIPTION
**Pull request by calippo**
Closes #26

- errors caused by distributed environment are poorly handled by autowire (e.g. timeouts etc), I'll open an issue about that
- Query is currently unimplemented
- context in `RPCClient` should take a list

usage recap:
```scala
val doghouseClient = deriveClientContext[DoghouseApi]
val rpcClient = new RPCClient(config, doghouseClient)

val res = rpcClient[DoghouseApi].getPuppy(1).call()
```

### tests performed
- tested locally

### CODE REVIEW

File: clientAkkaHttp/src/main/scala/RPCClient.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      )
+    )
+  }
+
+  override def doCall(autowireRequest: Request): Future[Json] = {
+    val completePath = autowireRequest.path.mkString(".")
+    //we're trying to match here the paths generated by two different macros
+    //if it fails at runtime it means something is wrong in the implementation
+    val methodMetaData = ctx.methodsMetaData
+      .getOrElse(completePath, throw new Exception("Runtime mismatch between metadata and method name"))
```
Comment (bytecodeguru) [inline clientAkkaHttp/src/main/scala/RPCClient.scala]: I'dd add more context to the error message: at least the method name.

File: clientAkkaHttp/src/main/scala/RPCClient.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    val operationName = methodMetaData.operationType.name.getOrElse(autowireRequest.path.last)
+    val uri = s"http://${config.host}:${config.port}/${ctx.path}/$operationName"
```
Comment (bytecodeguru) [inline clientAkkaHttp/src/main/scala/RPCClient.scala]: Why is the protocol `http` hardcoded here?

File: core/src/main/scala/models.scala [hunk 3]
```diff
@@ -3,12 +3,7 @@ package wiro
...
 package object models {
-  case class ClientConfig(
-    host: String,
-    port: Int
-  )
-
-  case class ServerConfig(
+  case class Config(
     host: String,
     port: Int
```
Comment (bytecodeguru) [inline core/src/main/scala/models.scala:8]: I'd add `protocol` or at least the possibility to choose between `http` and `https`.

File: clientAkkaHttp/src/main/scala/RPCClient.scala [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  ctx: RPCClientContext[_]
+)(implicit
+  system: ActorSystem,
+  materializer: ActorMaterializer,
+  executionContext: ExecutionContext
+) extends autowire.Client[Json, Decoder, Encoder] {
+  def write[Result: Encoder](r: Result): Json = r.asJson
+
+  def read[Result: Decoder](p: Json): Result = {
+    //This trick is required to match the Result type of autowire
```
Comment (bytecodeguru) [inline clientAkkaHttp/src/main/scala/RPCClient.scala:36]: Perhaps it would be better to move this code together with its  server side counterpart.

**Review by bytecodeguru — CHANGES_REQUESTED** [review]
Some minor suggestions, let's at least improve the error message. The rest LGTM 👍

File: core/src/main/scala/models.scala (hunk 3 above)
Comment (calippo) [reply core/src/main/scala/models.scala:8]: :+1:

Comment (calippo) [reply core/src/main/scala/models.scala:8]: hm.. need to think more about this, I'll open a https support issue

File: clientAkkaHttp/src/main/scala/RPCClient.scala (hunk 4 above)
Comment (calippo) [reply clientAkkaHttp/src/main/scala/RPCClient.scala:36]: :+1: opening issue

Comment (calippo) [reply clientAkkaHttp/src/main/scala/RPCClient.scala:36]: https://github.com/buildo/wiro/issues/46

File: core/src/main/scala/models.scala (hunk 3 above)
Comment (calippo) [reply core/src/main/scala/models.scala:8]: https://github.com/buildo/wiro/issues/45

File: clientAkkaHttp/src/main/scala/RPCClient.scala (hunk 2 above)
Comment (calippo) [reply clientAkkaHttp/src/main/scala/RPCClient.scala]: https://github.com/buildo/wiro/issues/45

**Comment by calippo** [comment]
@bytecodeguru improved exception message and opened appropriate issues

**Review by bytecodeguru — APPROVED** [review]
👍

**Pull request merged**